File: mobile-bridge/REST/BridgeServer.swift
import Foundation
import Network

@MainActor
final class BridgeServer: NSObject {
    static let shared = BridgeServer()

    // MARK: - Subsystems

    private(set) var wsServer: WSServer?
    private(set) var restServer: RESTServer?
    private(set) var discovery = BridgeDiscovery()
    private(set) var pairing = BridgePairingManager()

    // MARK: - Dependencies

    private(set) var appModel: AppModel?
    private(set) var isRunning = false
    private var observer: BridgeObserver?

    // MARK: - Lifecycle

    func start(appModel: AppModel) {
        guard !isRunning else { return }
        self.appModel = appModel

        wsServer = WSServer(port: 8765, bridge: self)
        wsServer?.start()

        restServer = RESTServer(port: 8766, bridge: self)
        restServer?.start()

        discovery.start(port: 8765) { _ in }

        observer = BridgeObserver(appModel: appModel, bridge: self)
        observer?.start()

        isRunning = true
        NSLog("Bridge: started")
    }

    func stop() {
        observer?.stop()
        wsServer?.stop()
        restServer?.stop()
        discovery.stop()
        pairing.revoke()
        appModel = nil
        isRunning = false
        NSLog("Bridge: stopped")
    }

    // MARK: - Public API

    var connectedClients: Int { wsServer?.clientCount ?? 0 }

    func newPairingCode() -> String {
        pairing.generateCode()
    }

    // MARK: - Outbound

    func broadcast(_ event: BridgeEvent) {
        wsServer?.broadcast(event)
    }

    func broadcastToThread(_ threadID: UUID, _ event: BridgeEvent) {
        wsServer?.broadcast(event)
    }

    // MARK: - Inbound messages

    func handleClientMessage(_ event: BridgeEvent) {
        guard let appModel else { return }
        switch event {
        case .pair(let code):
            handlePair(code)
        case .sendMessage(let threadID, let content):
            appModel.submitPrompt(threadID, content: content)
        case .approve(let threadID, let actionID):
            appModel.approveAction(threadID, actionID: actionID)
        case .reject(let threadID, let actionID):
            appModel.rejectAction(threadID, actionID: actionID)
        case .createThread(let projectID, let prompt):
            appModel.createThread(projectID: projectID, initialPrompt: prompt)
        case .switchModel(let threadID, let model):
            appModel.switchModel(threadID, to: model)
        default:
            break
        }
    }

    private func handlePair(_ code: String) {
        guard let token = pairing.validateAndIssueToken(code) else {
            broadcast(BridgeEvent.pairFailed(reason: "Invalid or expired code"))
            return
        }
        broadcast(BridgeEvent.paired(sessionToken: token))
    }

    // MARK: - REST data providers (used by RESTServer)

    func serveStatus() -> Data {
        guard let appModel else { return jsonResponse(["error": "not ready"], status: 503) }
        let info: [String: Any] = [
            "status": "ok",
            "version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown",
            "threadCount": appModel.threads.count,
            "paired": pairing.currentCode != nil,
            "connectedClients": connectedClients,
        ]
        return jsonResponse(info)
    }

    func serveThreadList() -> Data {
        guard let appModel else { return jsonResponse(["error": "not ready"], status: 503) }

        let summaries: [[String: Any]] = appModel.threads.map { thread in
            let project = appModel.project(thread.projectID)
            let preview = latestPreview(for: thread.id)
            return [
                "id": thread.id.uuidString,
                "projectID": thread.projectID.uuidString,
                "projectName": project?.name ?? "Unknown",
                "title": thread.title,
                "provider": thread.provider.rawValue,
                "model": thread.model ?? "",
                "lastStatus": thread.lastStatus.map(\.rawValue) ?? "",
                "hasUnread": thread.hasUnread,
                "updatedAt": ISO8601DateFormatter().string(from: thread.updatedAt),
                "messageCount": timelineCount(for: thread.id),
                "lastMessagePreview": preview ?? "",
            ]
        }
        return jsonResponse(summaries)
    }

    func serveThreadDetail(_ threadID: UUID) -> Data {
        guard let appModel else { return jsonResponse(["error": "not ready"], status: 503) }
        guard let thread = appModel.thread(threadID) else {
            return jsonResponse(["error": "thread not found"], status: 404)
        }
        let project = appModel.project(thread.projectID)
        let document = Storage.loadDocument(threadID)

        let entries: [[String: Any]] = document.timeline.map { item in
            var dict: [String: Any] = [
                "id": item.id,
                "kind": kindString(from: item.content),
                "date": ISO8601DateFormatter().string(from: item.date),
            ]
            switch item.content {
            case .user(let msg):
                dict["text"] = msg.text
            case .assistant(let msg):
                dict["text"] = msg.text
            case .reasoning(let block):
                dict["text"] = block.text
            case .tool(let call):
                dict["name"] = call.toolName
                dict["input"] = call.input ?? ""
                dict["output"] = call.output ?? ""
                dict["error"] = call.error ?? ""
            case .plan(let plan):
                dict["text"] = plan.markdown
            case .todos(let steps):
                dict["text"] = steps.map { "\($0.isDone ? "✓" : "○") \($0.title)" }.joined(separator: "\n")
            case .notice(let notice):
                dict["text"] = notice.text
            case .turnEnd(let summary):
                dict["summary"] = summary?.summary ?? ""
            }
            return dict
        }

        let detail: [String: Any] = [
            "id": thread.id.uuidString,
            "projectID": thread.projectID.uuidString,
            "projectPath": project?.path ?? "",
            "title": thread.title,
            "provider": thread.provider.rawValue,
            "model": thread.model ?? "",
            "runtimeMode": thread.runtimeMode.rawValue,
            "createdAt": ISO8601DateFormatter().string(from: thread.createdAt),
            "updatedAt": ISO8601DateFormatter().string(from: thread.updatedAt),
            "entries": entries,
            "approvals": [] as [[String: Any]],
            "diffRevision": 0,
        ]
        return jsonResponse(detail)
    }

    func serveModels() -> Data {
        let models: [String: Any] = [
            "providers": [
                "claude": ["id": "claude", "models": ["claude-sonnet-4", "claude-opus-4"]],
                "codex": ["id": "codex", "models": ["gpt-4o", "o1-preview"]],
                "deepseek": ["id": "deepseek", "models": ["deepseek-chat"]],
            ]
        ]
        return jsonResponse(models)
    }

    // MARK: - Private helpers

    private func kindString(from content: TimelineItem.Content) -> String {
        switch content {
        case .user: return "user"
        case .assistant: return "assistant"
        case .reasoning: return "reasoning"
        case .tool: return "tool"
        case .plan: return "plan"
        case .todos: return "todos"
        case .notice: return "notice"
        case .turnEnd: return "turnEnd"
        }
    }

    private func latestPreview(for threadID: UUID) -> String? {
        let doc = Storage.loadDocument(threadID)
        for item in doc.timeline.reversed() {
            switch item.content {
            case .user(let msg): return String(msg.text.prefix(80))
            case .assistant(let msg): return String(msg.text.prefix(80))
            case .tool(let call): return "[\(call.toolName)]"
            default: continue
            }
        }
        return nil
    }

    private func timelineCount(for threadID: UUID) -> Int {
        Storage.loadDocument(threadID).timeline.count
    }

    private func jsonResponse(_ value: Any, status: Int = 200) -> Data {
        guard let data = try? JSONSerialization.data(withJSONObject: value) else {
            return Data(#"{"error":"encoding failed"}"#.utf8)
        }
        var response = "HTTP/1.1 \(status == 200 ? "OK" : "Not Found")\r\n"
        response += "Content-Type: application/json\r\n"
        response += "Content-Length: \(data.count)\r\n"
        response += "Access-Control-Allow-Origin: *\r\n"
        response += "Connection: close\r\n"
        response += "\r\n"
        var packet = Data(response.utf8)
        packet.append(data)
        return packet
    }
}

File: scripts/bridge_server.mjs
import http from 'http';
import { WebSocketServer } from 'ws';

const PORT = 8765;
const PAIRING_CODE = '123456';

const sampleThreadId = 'a1b2c3d4-e5f6-4a5b-8c9d-0e1f2a3b4c5d';
const sampleProjectId = '00000000-0000-0000-0000-000000000000';

const sampleThreads = [
  {
    id: sampleThreadId,
    projectID: sampleProjectId,
    projectName: 'Swarm Code',
    title: 'Mobile Companion Setup',
    provider: 'anthropic',
    model: 'claude-3-7-sonnet',
    lastStatus: 'completed',
    hasUnread: false,
    updatedAt: new Date().toISOString(),
    messageCount: 2,
    lastMessagePreview: 'Swarm Code Mobile is now live and paired with your Mac.'
  }
];

const sampleThreadDetail = {
  id: sampleThreadId,
  projectID: sampleProjectId,
  projectPath: '/Users/soumyachakraborty/Documents/Swarm-Code',
  title: 'Mobile Companion Setup',
  provider: 'anthropic',
  model: 'claude-3-7-sonnet',
  runtimeMode: 'local',
  createdAt: new Date(Date.now() - 3600000).toISOString(),
  updatedAt: new Date().toISOString(),
  diffRevision: 1,
  approvals: [],
  entries: [
    {
      id: 'entry-1',
      kind: 'user',
      date: new Date(Date.now() - 3600000).toISOString(),
      text: 'Connect Swarm Code Mobile companion app to my Mac',
      name: null,
      input: null,
      output: null,
      error: null,
      summary: null
    },
    {
      id: 'entry-2',
      kind: 'assistant',
      date: new Date().toISOString(),
      text: 'Swarm Code Mobile is connected and communicating with your Mac over WebSocket. You can send messages, inspect threads, and manage coding agents from your phone!',
      name: null,
      input: null,
      output: null,
      error: null,
      summary: null
    }
  ]
};

const server = http.createServer((req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.writeHead(200);
    res.end();
    return;
  }

  const url = req.url || '';

  if (url === '/api/v1/status') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
      status: 'ok',
      version: '1.1.10',
      threadCount: sampleThreads.length,
      paired: true,
      connectedClients: wss.clients.size
    }));
    return;
  }

  if (url === '/api/v1/threads') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(sampleThreads));
    return;
  }

  if (url.startsWith('/api/v1/threads/')) {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(sampleThreadDetail));
    return;
  }

  if (url === '/api/v1/models') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
      providers: {
        anthropic: {
          id: 'anthropic',
          name: 'Anthropic Claude',
          models: ['claude-3-7-sonnet', 'claude-3-5-sonnet', 'claude-3-5-haiku']
        },
        openai: {
          id: 'openai',
          name: 'OpenAI',
          models: ['gpt-4o', 'o3-mini', 'gpt-4.5-preview']
        }
      }
    }));
    return;
  }

  res.writeHead(404, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ error: 'Not found' }));
});

const wss = new WebSocketServer({ server, path: '/ws/v1/stream' });

wss.on('connection', (ws) => {
  console.log('✓ Swarm Code Mobile connected over WebSocket!');

  ws.on('message', (data) => {
    try {
      const msg = JSON.parse(data.toString());
      console.log('Received event from mobile:', msg);

      if (msg.type === 'pair') {
        // Accept pairing code
        console.log(`Pairing code received: ${msg.code}. Pairing successful!`);
        ws.send(JSON.stringify({
          type: 'paired',
          sessionToken: 'swarm_session_' + Date.now()
        }));
      } else if (msg.type === 'sendMessage') {
        // Echo reply streaming
        const threadID = msg.threadID || sampleThreadId;
        setTimeout(() => {
          ws.send(JSON.stringify({
            type: 'messageDelta',
            threadID: threadID,
            content: `Echo from Swarm Code: "${msg.content}"`,
            isTool: false
          }));
        }, 300);

        setTimeout(() => {
          ws.send(JSON.stringify({
            type: 'messageComplete',
            threadID: threadID,
            messageID: 'msg_' + Date.now()
          }));
        }, 800);
      }
    } catch (err) {
      console.error('Error handling WS message:', err);
    }
  });

  ws.on('close', () => {
    console.log('Mobile client disconnected');
  });
});

server.listen(PORT, '0.0.0.0', () => {
  console.log(`=========================================`);
  console.log(`  Swarm Code Bridge Server LIVE on port ${PORT}`);
  console.log(`  Pairing Code: ${PAIRING_CODE} (any 6 digits will work)`);
  console.log(`=========================================`);
});
